File: wk6/java/removeOutermostParentheses.java
// problem link: https://leetcode.com/problems/remove-outermost-parentheses/

class Solution {
    public String removeOuterParentheses(String s) {
        StringBuilder ret = new StringBuilder();
        boolean prim = false;
        int c = 0;
        for(int i = 0; i < s.length(); i++) {
            if(s.charAt(i) == '(') {
                if(!prim)
                    prim = true;
                else {
                    ret.append('(');
                    c++;
                }
            }
            else {
                if(c == 0 && prim)
                    prim = false;
                else {
                    ret.append(')');
                    c--;
                }
            }
        }
        return ret.toString();
    }
}

File: wk6/cpp/removeOutermostParentheses.cpp
// problem link: https://leetcode.com/problems/remove-outermost-parentheses/description/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string removeOuterParentheses(string s) {
        int l = 0, c = 0;
        for(int r = 0; r < s.length(); r++) {
            if(c == 0)
                c++;
            else if(c == 1 && s[r] == ')')
                c--;
            else {
                c += (s[r] == '(' ? 1 : -1);
                s[l++] = s[r];
            }
        }
        s.resize(l);
        return s;
    }
};